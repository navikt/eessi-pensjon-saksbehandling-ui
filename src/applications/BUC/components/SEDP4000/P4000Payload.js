import _ from 'lodash'
import moment from 'moment'

export default class P4000Payload {
  constructor (pinfo, t) {
    this.pinfo = pinfo
    this.t = t
  }

  pinfoDateToDate (date) {
    if (!date || !date.year || !date.month) { return null }
    const day = parseInt(date.day) || 1
    const month = parseInt(date.month, 10) - 1
    const year = parseInt(date.year)
    return { year, month, day }
  }

  renderDate (date) {
    if (!date) {
      return this.t('ui:unknown')
    }
    const d = moment(this.pinfoDateToDate(date))
    return d.isValid() ? d.format('YYYY-MM-DD') : null
  }

  handleDate (period) {
    switch (period.dateType) {
      case 'both':
        return {
          lukketPeriode: {
            fom: this.renderDate(period.startDate),
            tom: this.renderDate(period.endDate)
          }
        }
      case 'onlyStartDate01':
        return {
          openPeriode: {
            extra: '01',
            fom: this.renderDate(period.startDate)
          }
        }
      case 'onlyStartDate98':
        return {
          openPeriode: {
            extra: '98',
            fom: this.renderDate(period.startDate)
          }
        }

      default:
        return null
    }
  }

  handleCountry (country) {
    return country.value
  }

  handleGenericPeriod (period) {
    const result = {}

    result.land = this.handleCountry(period.country)
    result.periode = this.handleDate(period)
    if (period.attachments) {
      result.vedlegg = period.attachments
    }
    if (period.insuranceName) {
      result.trygdeordningnavn = period.insuranceName
    }

    if (period.insuranceType) {
      result.medlemskap = period.insuranceType
    }

    if (period.insuranceId) {
      result.forsikkringEllerRegistreringNr = period.insuranceId
    }

    result.annenInformasjon = period.comment
    result.usikkerDatoIndikator = period.uncertainDate ? '1' : '0'
    return result
  }

  handleWorkPeriod (period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.jobbUnderAnsattEllerSelvstendig = period.workActivity
    newPeriod.navnFirma = period.workName
    newPeriod.typePeriode = period.workType
    newPeriod.adresseFirma = {
      postnummer: period.workZipCode,
      by: period.workCity,
      land: this.handleCountry(period.country),
      gate: period.workStreet,
      region: period.workRegion
    }
    delete newPeriod.land
    return newPeriod
  }

  handleDailyPeriod (period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.navnPaaInstitusjon = period.payingInstitution
    return newPeriod
  }

  handleSickPeriod (period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.navnPaaInstitusjon = period.payingInstitution
    return newPeriod
  }

  handleHomePeriod (period) {
    return this.handleGenericPeriod(period)
  }

  handleBirthPeriod (period) {
    return this.handleGenericPeriod(period)
  }

  handleVoluntaryPeriod (period) {
    return this.handleGenericPeriod(period)
  }

  handleMilitaryPeriod (period) {
    return this.handleGenericPeriod(period)
  }

  handleOtherPeriod (period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.typePeriode = period.otherType
    return newPeriod
  }

  handleChildPeriod (period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.informasjonBarn = {
      etternavn: period.childLastName,
      foedseldato: this.renderDate(period.childBirthDate),
      fornavn: period.childFirstName,
      land: this.handleCountry(period.country)
    }
    delete newPeriod.land
    return newPeriod
  }

  handleLearnPeriod (period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.navnPaaInstitusjon = period.learnInstitution
    return newPeriod
  }

  generatePayload () {
    return { periodeInfo: this.generatePeriods() }
  }

  generatePeriods () {
    const payload = {}
    this.pinfo.map(period => {
      switch (period.type) {
        case 'work':
          !_.has(payload, 'ansattSelvstendigPerioder')
            ? payload.ansattSelvstendigPerioder = [this.handleWorkPeriod(period)]
            : payload.ansattSelvstendigPerioder.push(this.handleWorkPeriod(period))
          break
        case 'home':
          !_.has(payload, 'boPerioder')
            ? payload.boPerioder = [this.handleHomePeriod(period)]
            : payload.boPerioder.push(this.handleHomePeriod(period))
          break
        case 'child':
          !_.has(payload, 'barnepassPerioder')
            ? payload.barnepassPerioder = [this.handleChildPeriod(period)]
            : payload.barnepassPerioder.push(this.handleChildPeriod(period))
          break
        case 'voluntary':
          !_.has(payload, 'frivilligPerioder')
            ? payload.frivilligPerioder = [this.handleVoluntaryPeriod(period)]
            : payload.frivilligPerioder.push(this.handleVoluntaryPeriod(period))
          break
        case 'military':
          !_.has(payload, 'forsvartjenestePerioder')
            ? payload.forsvartjenestePerioder = [this.handleMilitaryPeriod(period)]
            : payload.forsvartjenestePerioder.push(this.handleMilitaryPeriod(period))
          break
        case 'birth':
          !_.has(payload, 'foedselspermisjonPerioder')
            ? payload.foedselspermisjonPerioder = [this.handleBirthPeriod(period)]
            : payload.foedselspermisjonPerioder.push(this.handleBirthPeriod(period))
          break
        case 'learn':
          !_.has(payload, 'opplaeringPerioder')
            ? payload.opplaeringPerioder = [this.handleLearnPeriod(period)]
            : payload.opplaeringPerioder.push(this.handleLearnPeriod(period))
          break
        case 'daily':
          !_.has(payload, 'arbeidsledigPerioder')
            ? payload.arbeidsledigPerioder = [this.handleDailyPeriod(period)]
            : payload.arbeidsledigPerioder.push(this.handleDailyPeriod(period))
          break
        case 'sick':
          !_.has(payload, 'sykePerioder')
            ? payload.sykePerioder = [this.handleSickPeriod(period)]
            : payload.sykePerioder.push(this.handleSickPeriod(period))
          break
        case 'other':
          !_.has(payload, 'andrePerioder')
            ? payload.andrePerioder = [this.handleOtherPeriod(period)]
            : payload.andrePerioder.push(this.handleOtherPeriod(period))
          break
        default:
          return {}
      }
      return period
    })
    return payload
  }
}
