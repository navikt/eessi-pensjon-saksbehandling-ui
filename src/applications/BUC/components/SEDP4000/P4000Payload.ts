import { Country, P4000PayloadInfo, Payload, PayloadPeriod, Period, PeriodDate } from 'declarations/period'
import { T } from 'declarations/types'
import _ from 'lodash'
import moment, { Moment } from 'moment'

export default class P4000Payload {
  public pinfo: Array<Period>
  private t: T
  constructor (pinfo: Array<Period>, t: T) {
    this.pinfo = pinfo
    this.t = t
  }

  static pinfoDateToDate (date: PeriodDate | null): Date | null {
    if (!date || !date.year || !date.month) { return null }
    const day = typeof date.day === 'number' ? date.day : parseInt(date.day, 10) || 1
    const month = typeof date.month === 'number' ? date.month - 1 : parseInt(date.month, 10) - 1
    const year = typeof date.year === 'number' ? date.year : parseInt(date.year, 10)
    return new Date(year, month, day)
  }

  static renderDate (date: PeriodDate | null | undefined, t: T): string | null {
    if (!date) {
      return t('ui:unknown')
    } else {
      const dd: Date | null = P4000Payload.pinfoDateToDate(date)
      if (dd) {
        const d: Moment = moment(dd)
        return d.isValid() ? d.format('DD.MM.YYYY') : null
      } else {
        return null
      }
    }
  }

  renderDate (date: PeriodDate | null | undefined): string | null {
    return P4000Payload.renderDate(date, this.t)
  }

  handleDate (period: Period) {
    switch (period.dateType) {
      case 'both':
        return {
          lukketPeriode: {
            fom: this.renderDate(period.startDate),
            tom: this.renderDate(period.endDate)
          }
        }
      case 'onlyStartDate01':
        return {
          openPeriode: {
            extra: '01',
            fom: this.renderDate(period.startDate)
          }
        }
      case 'onlyStartDate98':
        return {
          openPeriode: {
            extra: '98',
            fom: this.renderDate(period.startDate)
          }
        }

      default:
        return null
    }
  }

  handleCountry (country: Country): string {
    return country.value
  }

  handleGenericPeriod (period: Period) {
    const result: PayloadPeriod = {
      land: this.handleCountry(period.country),
      periode: this.handleDate(period)
    }

    if (period.attachments) {
      result.vedlegg = period.attachments
    }
    if (period.insuranceName) {
      result.trygdeordningnavn = period.insuranceName
    }

    if (period.insuranceType) {
      result.medlemskap = period.insuranceType
    }

    if (period.insuranceId) {
      result.forsikkringEllerRegistreringNr = period.insuranceId
    }

    result.annenInformasjon = period.comment
    result.usikkerDatoIndikator = period.uncertainDate ? '1' : '0'
    return result
  }

  handleWorkPeriod (period: Period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.jobbUnderAnsattEllerSelvstendig = period.workActivity
    newPeriod.navnFirma = period.workName
    newPeriod.typePeriode = period.workType
    newPeriod.adresseFirma = {
      postnummer: period.workZipCode,
      by: period.workCity,
      land: this.handleCountry(period.country),
      gate: period.workStreet,
      region: period.workRegion
    }
    delete newPeriod.land
    return newPeriod
  }

  handleDailyPeriod (period: Period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.navnPaaInstitusjon = period.payingInstitution
    return newPeriod
  }

  handleSickPeriod (period: Period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.navnPaaInstitusjon = period.payingInstitution
    return newPeriod
  }

  handleHomePeriod (period: Period) {
    return this.handleGenericPeriod(period)
  }

  handleBirthPeriod (period: Period) {
    return this.handleGenericPeriod(period)
  }

  handleVoluntaryPeriod (period: Period) {
    return this.handleGenericPeriod(period)
  }

  handleMilitaryPeriod (period: Period) {
    return this.handleGenericPeriod(period)
  }

  handleOtherPeriod (period: Period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.typePeriode = period.otherType
    return newPeriod
  }

  handleChildPeriod (period: Period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.informasjonBarn = {
      etternavn: period.childLastName,
      foedseldato: this.renderDate(period.childBirthDate),
      fornavn: period.childFirstName,
      land: this.handleCountry(period.country)
    }
    delete newPeriod.land
    return newPeriod
  }

  handleLearnPeriod (period: Period) {
    const newPeriod = this.handleGenericPeriod(period)
    newPeriod.navnPaaInstitusjon = period.learnInstitution
    return newPeriod
  }

  generatePayload (): P4000PayloadInfo {
    return { periodeInfo: this.generatePeriods() } as P4000PayloadInfo
  }

  generatePeriods (): Payload {
    const payload: Payload = {}
    this.pinfo.map((period: Period) => {
      switch (period.type) {
        case 'work':
          !_.has(payload, 'ansattSelvstendigPerioder')
            ? payload.ansattSelvstendigPerioder = [this.handleWorkPeriod(period)]
            : payload.ansattSelvstendigPerioder!.push(this.handleWorkPeriod(period))
          break
        case 'home':
          !_.has(payload, 'boPerioder')
            ? payload.boPerioder = [this.handleHomePeriod(period)]
            : payload.boPerioder!.push(this.handleHomePeriod(period))
          break
        case 'child':
          !_.has(payload, 'barnepassPerioder')
            ? payload.barnepassPerioder = [this.handleChildPeriod(period)]
            : payload.barnepassPerioder!.push(this.handleChildPeriod(period))
          break
        case 'voluntary':
          !_.has(payload, 'frivilligPerioder')
            ? payload.frivilligPerioder = [this.handleVoluntaryPeriod(period)]
            : payload.frivilligPerioder!.push(this.handleVoluntaryPeriod(period))
          break
        case 'military':
          !_.has(payload, 'forsvartjenestePerioder')
            ? payload.forsvartjenestePerioder = [this.handleMilitaryPeriod(period)]
            : payload.forsvartjenestePerioder!.push(this.handleMilitaryPeriod(period))
          break
        case 'birth':
          !_.has(payload, 'foedselspermisjonPerioder')
            ? payload.foedselspermisjonPerioder = [this.handleBirthPeriod(period)]
            : payload.foedselspermisjonPerioder!.push(this.handleBirthPeriod(period))
          break
        case 'learn':
          !_.has(payload, 'opplaeringPerioder')
            ? payload.opplaeringPerioder = [this.handleLearnPeriod(period)]
            : payload.opplaeringPerioder!.push(this.handleLearnPeriod(period))
          break
        case 'daily':
          !_.has(payload, 'arbeidsledigPerioder')
            ? payload.arbeidsledigPerioder = [this.handleDailyPeriod(period)]
            : payload.arbeidsledigPerioder!.push(this.handleDailyPeriod(period))
          break
        case 'sick':
          !_.has(payload, 'sykePerioder')
            ? payload.sykePerioder = [this.handleSickPeriod(period)]
            : payload.sykePerioder!.push(this.handleSickPeriod(period))
          break
        case 'other':
          !_.has(payload, 'andrePerioder')
            ? payload.andrePerioder = [this.handleOtherPeriod(period)]
            : payload.andrePerioder!.push(this.handleOtherPeriod(period))
          break
        default:
          return {}
      }
      return period
    })
    return payload
  }
}
