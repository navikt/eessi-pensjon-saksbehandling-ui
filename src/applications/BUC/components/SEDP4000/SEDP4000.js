import React, { useState, useEffect } from 'react'
import PT from 'prop-types'
import classNames from 'classnames'
import _ from 'lodash'
import { connect } from 'store'
import { NavFrontendSpinner, Undertekst, Undertittel } from 'components/Nav'
import Period from './Period'

export const mapStateToProps = (state) => {
  return {
    aktoerId: state.app.params.aktoerId,
    p4000: state.buc.p4000,
    locale: state.ui.locale,
    loadingP4000list: state.loading.loadingP4000list,
    loadingP4000info: state.loading.loadingP4000info,
    p4000list: state.buc.p4000list,
    p4000info: state.buc.p4000info
  }
}

export const SEDP4000 = (props) => {
  const [ period, setPeriod ] = useState({})
  const [ maxPeriods ] = useState(8)
  const [ isReady, setIsReady ] = useState(false)

  const { actions, aktoerId, loadingP4000info, loadingP4000list, locale } = props
  const { p4000info, p4000list, setShowButtons, showButtons, t } = props
  const mode = period && period.id ? 'edit' : 'new'

  // check aktoerId
  useEffect(() => {
    if (!aktoerId) {
      setIsReady(true)
    }
  }, [aktoerId])

  // load P4000 file list
  useEffect(() => {
    if (!isReady && !loadingP4000list && p4000list === undefined) {
      actions.listP4000(aktoerId)
    }
  }, [aktoerId, actions, isReady, loadingP4000list, p4000list])

  // load P4000 info file if exists
  useEffect(() => {
    if (!isReady && !loadingP4000info && p4000info === undefined && _.isArray(p4000list)) {
      const p4000file = aktoerId + '___PINFO___PINFO.json'
      if (p4000list.indexOf(p4000file) >= 0) {
        actions.getP4000(p4000file)
      } else {
        setIsReady(true)
      }
    }
  }, [actions, aktoerId, loadingP4000info, p4000info, p4000list, isReady])

  // mark component as ready if p4000 info file is ready
  useEffect(() => {
    if (!isReady && p4000info !== undefined) {
      setIsReady(true)
    }
  }, [isReady, p4000info])

  useEffect(() => {
    // we are editing a period or adding a new period => hide main buttons
    if ((period.type || period.id) && showButtons) {
      setShowButtons(false)
    }
    // we are NOT editing a period or adding a new period => show main buttons
    if (!period.type && !period.id && !showButtons) {
      setShowButtons(true)
    }
  }, [mode, setShowButtons, showButtons, period])

  const setPeriods = (periods) => {
    actions.setP4000Info({
      ...p4000info,
      stayAbroad: periods
    })
  }

  if (!aktoerId) {
    return <div>{t('buc:validation-noAktoerId')}</div>
  }

  if (!isReady) {
    return <div className='col-md-8 d-flex mt-5 mb-5 align-items-center justify-content-center'>
      <NavFrontendSpinner className='ml-3 mr-3' type='M' />
      { loadingP4000list ? <span className='pl-2'>{t('buc:loading-p4000list')}</span> : null }
      { loadingP4000info ? <span className='pl-2'>{t('buc:loading-p4000info')}</span> : null }
    </div>
  }

  return <div className={classNames('a-buc-c-sedp4000', mode)}>
    {mode === 'new' ? <React.Fragment>
      <Undertittel className='mb-3'>{t('buc:p4000-title')}</Undertittel>
      <Undertekst className='mb-2'>{t('buc:p4000-description')}</Undertekst>
      <Undertekst className='mb-3'>{t('buc:p4000-info-on-help-icon')}</Undertekst>
    </React.Fragment> : null}
    {!_.isEmpty(p4000info.stayAbroad) && mode === 'new' ? <React.Fragment>
      <Undertittel className='mt-5 mb-2'>{t('buc:p4000-previousPeriods')}</Undertittel>
      {p4000info.stayAbroad.sort((a, b) => {
        return a.startDate - b.startDate
      }).map((period, index) => {
        return <Period t={t}
          locale={locale}
          actions={actions}
          mode='view'
          first={index === 0}
          last={index === p4000info.stayAbroad.length - 1}
          period={period}
          periods={p4000info.stayAbroad}
          setPeriod={setPeriod}
          setPeriods={setPeriods}
          key={index} />
      })}
    </React.Fragment>
      : null}
    { p4000info.stayAbroad.length < maxPeriods ?
    <Period t={t}
      actions={actions}
      mode={mode}
      showButtons={showButtons}
      period={period}
      periods={p4000info.stayAbroad}
      locale={locale}
      setPeriod={setPeriod}
      setPeriods={setPeriods}
    /> :
    <span>
      {t('buc:p4000-alert-maxPeriods', { maxPeriods: maxPeriods })}
    </span> }
  </div>
}

SEDP4000.propTypes = {
  actions: PT.object,
  aktoerId: PT.string.isRequired,
  loadingP4000info: PT.bool,
  loadingP4000list: PT.bool,
  locale: PT.string.isRequired,
  p4000info: PT.object,
  p4000list: PT.array,
  setShowButtons: PT.func.isRequired,
  showButtons: PT.bool.isRequired,
  t: PT.func
}

export default connect(mapStateToProps, () => {})(SEDP4000)
