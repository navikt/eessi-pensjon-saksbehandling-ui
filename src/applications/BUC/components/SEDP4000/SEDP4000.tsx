import Period from 'applications/BUC/components/SEDP4000/Period/Period'
import { P4000Info, Period as PeriodD, PeriodErrors } from 'applications/BUC/declarations/period'
import classNames from 'classnames'
import * as storage from 'constants/storage'
import { Nav } from 'eessi-pensjon-ui'
import _ from 'lodash'
import PT from 'prop-types'
import React, { useEffect, useState } from 'react'
import { connect } from 'store'
import { ActionCreators, AllowedLocaleString, State, T } from 'types'
import P4000Payload from './P4000Payload'

export interface SEDP4000Props {
  actions: ActionCreators;
  aktoerId: string;
  loadingP4000info: boolean;
  loadingP4000list: boolean;
  locale: AllowedLocaleString;
  p4000info: P4000Info | undefined;
  p4000list: Array<string> | null | undefined;
  setShowButtons: Function;
  showButtons: boolean;
  t: T
}

export const mapStateToProps: Function = /* istanbul ignore next */ (state: State): any => {
  return {
    aktoerId: state.app.params.aktoerId,
    p4000: state.buc.p4000,
    locale: state.ui.locale,
    loadingP4000list: state.loading.loadingP4000list,
    loadingP4000info: state.loading.loadingP4000info,
    p4000list: state.buc.p4000list,
    p4000info: state.buc.p4000info
  }
}

export const SEDP4000 = ({
  actions, aktoerId, loadingP4000info, loadingP4000list,
  locale, p4000info, p4000list, setShowButtons, showButtons, t
}: SEDP4000Props) => {
  const [period, setPeriod] = useState<PeriodD>({} as PeriodD)
  const [isReady, setIsReady] = useState<boolean>(false)
  const [localErrors, setLocalErrors] = useState<PeriodErrors>({})
  const [p4000file, setP4000file] = useState<string | undefined | null>(undefined)
  const [role, setRole] = useState<string | undefined>(undefined)
  const mode = period.id ? 'edit' : 'new'

  // check aktoerId
  useEffect(() => {
    if (!aktoerId) {
      setIsReady(true)
    }
  }, [aktoerId])

  // mark component as ready if p4000 info file is ready
  useEffect(() => {
    if (!isReady && p4000info !== undefined) {
      setIsReady(true)
    }
  }, [isReady, p4000info])

  useEffect(() => {
    // we are editing a period or adding a new period => hide main buttons
    if ((period.type || period.id) && showButtons) {
      setShowButtons(false)
    }
    // we are NOT editing a period or adding a new period => show main buttons
    if (!period.type && !period.id && !showButtons) {
      setShowButtons(true)
    }
  }, [mode, setShowButtons, showButtons, period])

  const setLocalError = (key: string, error: string | undefined): void => {
    setLocalErrors({
      ...localErrors,
      [key]: error
    })
  }

  const setPeriods = (periods: Array<PeriodD>): void => {
    actions.setP4000Info({
      ...p4000info,
      stayAbroad: periods
    })
  }

  const handleContinueButton: Function = (): void => {
    actions.saveP4000asSaksbehandler(aktoerId, p4000file)
    setIsReady(true)
  }

  const handleGetP4000infoButton: Function = (): void => {
    actions.getP4000(p4000file)
  }

  const handleListP4000userButton: Function = (): void => {
    setRole('user')
    setP4000file(aktoerId + '___' + storage.NAMESPACE_PINFO + '___' + storage.FILE_PINFO)
    actions.listP4000(aktoerId)
  }

  const handleListP4000saksbehandlerButton = () => {
    setRole('saksbehandler')
    setP4000file(aktoerId + '___' + storage.NAMESPACE_PINFO + '___' + storage.FILE_PINFOSB)
    actions.listP4000(aktoerId)
  }

  const noP4000Info = () => {
    return p4000list === null || (_.isArray(p4000list) && p4000file && p4000list.indexOf(p4000file) < 0)
  }

  if (!aktoerId) {
    return <div>{t('buc:validation-noAktoerId')}</div>
  }

  const errorMessage = () => {
    for (var key in localErrors) {
      if (localErrors[key]) {
        return localErrors[key]
      }
    }
    return undefined
  }

  const _errorMessage = errorMessage()

  if (!isReady) {
    return (
      <div className='a-buc-c-sedp4000__notReady col-md-8 d-flex mt-5 mb-5 align-items-center justify-content-center'>
        {loadingP4000list || loadingP4000info ? <Nav.Spinner className='ml-3 mr-3' type='M' /> : null}
        {loadingP4000list ? <span className='pl-2'>{t('buc:loading-p4000list')}</span> : null}
        {loadingP4000info ? <span className='pl-2'>{t('buc:loading-p4000info')}</span> : null}
        {!loadingP4000info && p4000list !== undefined && p4000info === undefined
          ? (noP4000Info() ? (
            <span>
              <Nav.Normaltekst>{t('buc:p4000-label-p4000' + role + 'FileNotFound')}</Nav.Normaltekst>
              <Nav.Hovedknapp
                id='a-buc-c-sedp4000__continue-button-id'
                className='a-buc-c-sedp4000__continue-button mt-3'
                onClick={handleContinueButton}
              >
                {t('ui:continue')}
              </Nav.Hovedknapp>
            </span>
          ) : (
            <span>
              <Nav.Normaltekst>{t('buc:p4000-label-p4000' + role + 'FileFound')}</Nav.Normaltekst>
              <Nav.Hovedknapp
                id='a-buc-c-sedp4000__getP4000info-button-id'
                className='a-buc-c-sedp4000__getP4000info-button mt-3'
                onClick={handleGetP4000infoButton}
              >
                {t('ui:getInfo')}
              </Nav.Hovedknapp>
            </span>)
          ) : null}
        {!loadingP4000list && p4000list === undefined ? (
          <div className='d-flex flex-column'>
            <div>
              <Nav.Normaltekst>{t('buc:p4000-help-listP4000user')}</Nav.Normaltekst>
              <Nav.Hovedknapp
                id='a-buc-c-sedp4000__listP4000user-button-id'
                className='a-buc-c-sedp4000__listP4000user-button mt-3'
                onClick={handleListP4000userButton}
              >
                {t('buc:p4000-button-listP4000user')}
              </Nav.Hovedknapp>
            </div>
            <div className='mt-3'>
              <Nav.Normaltekst>{t('buc:p4000-help-listP4000saksbehandler')}</Nav.Normaltekst>
              <Nav.Hovedknapp
                id='a-buc-c-sedp4000__listP4000saksbehandler-button-id'
                className='a-buc-c-sedp4000__listP4000saksbehandler-button mt-3'
                onClick={handleListP4000saksbehandlerButton}
              >
                {t('buc:p4000-button-listP4000saksbehandler')}
              </Nav.Hovedknapp>
            </div>
          </div>
        ) : null}
      </div>
    )
  }

  return (
    <div className={classNames('a-buc-c-sedp4000', mode)}>
      {mode === 'new' ? (
        <>
          <Nav.Undertittel className='mb-3'>{t('buc:p4000-app-title')}</Nav.Undertittel>
          <Nav.Undertekst className='mb-2'>{t('buc:p4000-app-description')}</Nav.Undertekst>
          <Nav.Undertekst className='mb-3'>{t('buc:p4000-app-help')}</Nav.Undertekst>
        </>
      ) : null}
      {_errorMessage ? (
        <Nav.AlertStripe className='a-buc-c-sedp4000d__alert mt-4 mb-4' type='advarsel'>
          {t(_errorMessage)}
        </Nav.AlertStripe>
      ) : null}
      {p4000info && !_.isEmpty(p4000info.stayAbroad) && mode === 'new' ? (
        <>
          <Nav.Undertittel className='mt-5 mb-2'>{t('buc:p4000-title-previousPeriods')}</Nav.Undertittel>
          {p4000info ? p4000info.stayAbroad.sort((a: PeriodD, b: PeriodD) => {
            return P4000Payload.pinfoDateToDate(a.startDate)!.getDate()  - P4000Payload.pinfoDateToDate(b.startDate)!.getDate()
          }).map((period, index) => {
            return (
              <Period
                t={t}
                locale={locale}
                actions={actions}
                mode='view'
                first={index === 0}
                last={index === p4000info.stayAbroad.length - 1}
                period={period}
                periods={p4000info.stayAbroad}
                localErrors={localErrors}
                setLocalError={setLocalError}
                setLocalErrors={setLocalErrors}
                setPeriod={setPeriod}
                setPeriods={setPeriods}
                key={index}
              />
            )
          }) : null}
        </>
      )
        : null}
      <Period
        t={t}
        actions={actions}
        mode={mode}
        period={period}
        periods={p4000info ? p4000info.stayAbroad : []}
        locale={locale}
        localErrors={localErrors}
        setLocalError={setLocalError}
        setLocalErrors={setLocalErrors}
        setPeriod={setPeriod}
        setPeriods={setPeriods}
      />
      {_errorMessage ? (
        <Nav.AlertStripe className='a-buc-c-sedp4000d__alert mt-4 mb-4' type='advarsel'>
          {t(_errorMessage)}
        </Nav.AlertStripe>
      ) : null}
    </div>
  )
}

SEDP4000.propTypes = {
  actions: PT.object,
  aktoerId: PT.string.isRequired,
  loadingP4000info: PT.bool,
  loadingP4000list: PT.bool,
  locale: PT.string.isRequired,
  p4000info: PT.object,
  p4000list: PT.array,
  setShowButtons: PT.func.isRequired,
  showButtons: PT.bool.isRequired,
  t: PT.func
}

export default connect(mapStateToProps, () => {})(SEDP4000)
