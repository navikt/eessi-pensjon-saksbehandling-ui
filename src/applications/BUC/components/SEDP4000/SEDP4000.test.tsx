import { listP4000 } from 'actions/buc'
import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import sampleP4000info from 'resources/tests/sampleP4000info'
import { SEDP4000, SEDP4000Props } from './SEDP4000'
import { useDispatch, useSelector } from 'react-redux'
jest.mock('react-redux')
jest.mock('./Period/Period', () => () => <div className='mock-period' />)
jest.mock('actions/buc', () => ({
  listP4000: jest.fn()
}));

(useDispatch as jest.Mock).mockImplementation(() => jest.fn())

const defaultSelector = {
  loadingP4000list: false,
  loadingP4000info: false,
  p4000info: undefined,
  p4000list: undefined
};

(useSelector as jest.Mock).mockImplementation(() => (defaultSelector))

describe('applications/BUC/components/SEDP4000/SEDP4000', () => {
  let wrapper: ReactWrapper
  const initialMockProps: SEDP4000Props = {
    aktoerId: '123',
    locale: 'nb',
    showButtons: true,
    setShowButtons: jest.fn(),
    t: jest.fn(t => t)
  }

  beforeEach(() => {
    wrapper = mount(<SEDP4000 {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Renders', () => {
    expect(wrapper.isEmptyRender()).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })

  it('Call for p4000 file list when clicking user-button', () => {
    wrapper.find('button#a-buc-c-sedp4000__listP4000user-button-id').simulate('click')
    expect(listP4000).toHaveBeenCalledWith(initialMockProps.aktoerId)
  })

  it('Call for p4000 file list when clicking saksbehandler-button', () => {
    wrapper.find('button#a-buc-c-sedp4000__listP4000saksbehandler-button-id').simulate('click')
    expect(listP4000).toHaveBeenCalledWith(initialMockProps.aktoerId)
  })

  /* it('Has proper HTML structure: unmounted state', () => {
    expect(wrapper.exists('.a-buc-c-sedp4000__notReady')).toBeTruthy()
    wrapper.setProps({ loadingP4000list: true })
    wrapper.update()
    expect(wrapper.exists('.a-buc-c-sedp4000__notReady Spinner')).toBeTruthy()
    expect(wrapper.find('.a-buc-c-sedp4000__notReady').render().text()).toEqual('Venter...buc:loading-p4000list')
    wrapper.setProps({ loadingP4000list: false, loadingP4000info: true })
    wrapper.update()
    expect(wrapper.exists('.a-buc-c-sedp4000__notReady Spinner')).toBeTruthy()
    expect(wrapper.find('.a-buc-c-sedp4000__notReady > span').text()).toEqual('buc:loading-p4000info')
  }) */

  it('Has proper HTML structure: mounted state', () => {
    (useSelector as jest.Mock).mockImplementation(() => ({
      ...defaultSelector,
      p4000list: ['sampleP4000file'],
      p4000info: sampleP4000info
    }))
    wrapper = mount(<SEDP4000 {...initialMockProps} />)
    expect(wrapper.exists('.a-buc-c-sedp4000')).toBeTruthy()
  })
})
