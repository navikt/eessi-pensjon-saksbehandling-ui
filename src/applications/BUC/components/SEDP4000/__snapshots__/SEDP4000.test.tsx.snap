// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications/BUC/components/SEDP4000/SEDP4000 Renders 1`] = `
<SEDP4000
  aktoerId="123"
  locale="nb"
  setShowButtons={[MockFunction]}
  showButtons={true}
>
  <div
    className="a-buc-c-sedp4000__notReady col-md-8 d-flex mt-5 mb-5 align-items-center justify-content-center"
  >
    <div
      className="d-flex flex-column"
    >
      <div>
        <Normaltekst>
          <n
            type="normaltekst"
          >
            <p
              className="typo-normal"
            >
              buc:p4000-help-listP4000user
            </p>
          </n>
        </Normaltekst>
        <Hovedknapp
          className="a-buc-c-sedp4000__listP4000user-button mt-3"
          id="a-buc-c-sedp4000__listP4000user-button-id"
          onClick={[Function]}
        >
          <KnappBase
            autoDisableVedSpinner={false}
            className="a-buc-c-sedp4000__listP4000user-button mt-3"
            disabled={false}
            htmlType="submit"
            id="a-buc-c-sedp4000__listP4000user-button-id"
            inaktivKlasseVedDisabled={false}
            kompakt={false}
            mini={false}
            onClick={[Function]}
            spinner={false}
            type="hoved"
          >
            <button
              className="knapp a-buc-c-sedp4000__listP4000user-button mt-3 knapp--hoved"
              disabled={false}
              id="a-buc-c-sedp4000__listP4000user-button-id"
              onClick={[Function]}
              type="submit"
            >
              buc:p4000-button-listP4000user
            </button>
          </KnappBase>
        </Hovedknapp>
      </div>
      <div
        className="mt-3"
      >
        <Normaltekst>
          <n
            type="normaltekst"
          >
            <p
              className="typo-normal"
            >
              buc:p4000-help-listP4000saksbehandler
            </p>
          </n>
        </Normaltekst>
        <Hovedknapp
          className="a-buc-c-sedp4000__listP4000saksbehandler-button mt-3"
          id="a-buc-c-sedp4000__listP4000saksbehandler-button-id"
          onClick={[Function]}
        >
          <KnappBase
            autoDisableVedSpinner={false}
            className="a-buc-c-sedp4000__listP4000saksbehandler-button mt-3"
            disabled={false}
            htmlType="submit"
            id="a-buc-c-sedp4000__listP4000saksbehandler-button-id"
            inaktivKlasseVedDisabled={false}
            kompakt={false}
            mini={false}
            onClick={[Function]}
            spinner={false}
            type="hoved"
          >
            <button
              className="knapp a-buc-c-sedp4000__listP4000saksbehandler-button mt-3 knapp--hoved"
              disabled={false}
              id="a-buc-c-sedp4000__listP4000saksbehandler-button-id"
              onClick={[Function]}
              type="submit"
            >
              buc:p4000-button-listP4000saksbehandler
            </button>
          </KnappBase>
        </Hovedknapp>
      </div>
    </div>
  </div>
</SEDP4000>
`;
